### ISSUE TITLE
Snackbar features

### ISSUE THREAD

**Issue by NeverGivinUp** [ZieIony/Carbon#84]
[Google's description of snackbars](http://www.google.com/design/spec/components/snackbars-toasts.html#) describes a number of features for their appearance.
1. disappear after a timeout, but there are missing features.
2. disappear after user interaction elsewhere on the screen
3. ability to swipe off the screen
4. moving away a floating action button (see video labeled "Animation of mobile snackbar", in Googl's description)

Of these features, Carbon's snackbar implements only the first. 

Hiding the snackbar on user interaction and the ability to swipe it off are quite self contained features. Moving away the floating action button requires a setter, so the snackbar knows where to find the button, it needs to move when animating in and needs to move back when animating out.

I think it is possible for apps, which use Carbon, to implement the features themselves. But it would certainly be nicer, if they didn't have to. Either way it would be good to have clarity which features Carbon will provide and which are the responsibility of the app.

**Comment by NeverGivinUp** [comment]
Thinking about it some more, moving the floating action button away should not even be a feature of the snackbar but a general feature of all material layouts: Any two elements, which are contained in a single layout and have the same elevation cannot overlap, in the real world of materials. If one is animated into the space of another, the other must be pushed away, as a function of the layout, which is wrapping them. This is what the pushing away of the floating action button is about, as --according to Google's description -- the floating action button and the snack bar have the same elevation.

**Comment by ZieIony** [comment]
I think that it may be very difficult to implement pushing other components away. I would like to have as simple interface as possible, ideally comparable with Toasts. It means that a Snackbar should attach itself to the root view or float in a Window. It means that both the Snackbar and the pushed view would be added to two separate layouts/windows. In such case it's very hard to properly react and layout both views.

Two thoughts:
- In Carbon drawing overlapping views may lead to unpredictable results, because it's impossible to determine the correct drawing order for two views with identical elevation value.
- In Material Design there are cases where two pieces of paper can merge, which could be implemented with two overlapping views.

**Comment by NeverGivinUp** [comment]
I can imagine that it would be difficult to correctly implement layouts, which enable same-elevated views to push each other away. It should be easier to do on a widget basis, though. And pushing away the floating action button by the snack bar is a material design requirement. 

To do it right, the pushing away needs to use the exact same interpolator and distance as the snack bar. The using app could do it, if the snack bar makes this information available via getter methods. The snackbar could apply its animation to a number of given views. A possible interface would be `mySnackbar.addPushedView(View v)` (and for the sake of completeness also `mySnackbar.removePushedView(View v)`). Or even simpler when constrained to a single view: ``mySnackbar.setPushedView(View v)`. The snack bar would animate all pushed views in parallel to its own in and out animation. It might log a warning, if they are not on the same elevation level as the snack bar.

What are your thoughts about the other two features of the snack bar (# 2 and 3 in the original post). Are you planning to implement them or do you think this should be the responsibility of the user?

**Comment by ZieIony** [comment]
Thanks for the ideas!

I would like to implement these two missing features (# 2 and 3).

**Comment by NeverGivinUp** [comment]
Sorry to bother you with yet another feature request:

The snack bar needs not just an onClickListener, which is called, when the user clicks the action button. It also needs a listener for when the Snackbar disappears -- e.g. it is swiped off the screen, times out or the user clicks somewhere else. 

For the rather common use case of a snack bar being used to offer an undo this enables an alternative implementation. In the first implementation the user signals to delete something, the app deletes it and shows the snackbar with the undo button. If the user clicks undo the app undeletes the thing. In the alternative the app does not delete what the user signaled to delete but just hides it. If the user hits undo, it just displays it again, but if the snack bar disappears without the user hitting undo, the app really deletes the thing.

**Comment by ZieIony** [comment]
Dismissing by swiping and by tapping outside is done. So is the dismiss listener. I've added a new issue for tracking pushing views. See #98
